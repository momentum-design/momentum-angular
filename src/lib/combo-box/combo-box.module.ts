import { OverlayModule } from '@angular/cdk/overlay';
import { CommonModule } from '@angular/common';
import { NgModule, CUSTOM_ELEMENTS_SCHEMA, ModuleWithProviders } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { InputModule } from '../input';
import { ListItemModule } from '../list-item';
import { ListItemHeaderModule } from '../list-item-header';
import { SearchInputModule } from '../search-input';
import { ComboBoxComponent } from './combo-box.component';

@NgModule({
  imports: [
    CommonModule,
    InputModule,
    ListItemModule,
    ListItemHeaderModule,
    OverlayModule,
    SearchInputModule,
    FormsModule,
    ReactiveFormsModule,
  ],
  declarations: [ComboBoxComponent],
  exports: [ComboBoxComponent],
  schemas: [CUSTOM_ELEMENTS_SCHEMA],
})
export class ComboBoxModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: ComboBoxModule,
      providers: [],
    };
  }
}
