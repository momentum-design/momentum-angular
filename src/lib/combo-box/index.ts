export * from './combo-box.module';
export * from './combo-box.component';
