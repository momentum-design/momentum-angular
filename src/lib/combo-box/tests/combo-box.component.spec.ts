import { Component, ViewChild, Type, Provider, } from '@angular/core';
import { fakeAsync, ComponentFixture, flush, TestBed, inject, tick } from '@angular/core/testing';
import { OverlayContainer } from '@angular/cdk/overlay';
import { MutationObserverFactory } from '@angular/cdk/observers';
import { FormsModule } from '@angular/forms';
import {
  ComboBoxModule,
} from 'src/lib/public_api';
import { ComboBoxComponent } from '../combo-box.component';

@Component({
  selector: 'test-combo-box',
  template: `
    <md-combo-box
      [options]="options"
      [clear]="clear"
      [value]="value"
      [(ngModel)]="value"
      [hasSearchIcon]="hasSearchIcon"
      [ngClass]="'test'"
      (select)="onSelect($event)"
      (change)="onChange($event)"
    >
    </md-combo-box>
  `
})
class TestComboBoxComponent {
  @ViewChild(ComboBoxComponent) comboBoxComponent: ComboBoxComponent;

  clear = false;
  hasSearchIcon = true;
  value = '';
  options: Array<Object | string> = ['a', 'ab', 'abc'];

  onSelect(option: Object | string) {}
  onChange(event: Event) {}
}

describe('ComboBoxComponent', () => {
  let testComponent: TestComboBoxComponent;
  let fixture: ComponentFixture<TestComboBoxComponent>;
  let nativeElement: HTMLElement;
  let overlayContainer: OverlayContainer;
  let overlayContainerElement: HTMLElement;

  function createComponent<T>(component: Type<T>,
                              declarations: any[] = []): ComponentFixture<T> {
    TestBed.configureTestingModule({
      imports: [
        ComboBoxModule,
        FormsModule,
      ],
      declarations: [component, ...declarations],
      providers: [{
        provide: MutationObserverFactory,
        useValue: {
          create: (callback: Function) => {
            return {
              observe: () => {},
              disconnect: () => {}
            };
          }
        }
      }]
    }).compileComponents();

    inject([OverlayContainer], (oc: OverlayContainer) => {
      overlayContainer = oc;
      overlayContainerElement = oc.getContainerElement();
    })();

    return TestBed.createComponent<T>(component);
  }

  beforeEach(() => {
    fixture = createComponent(TestComboBoxComponent);
    testComponent = fixture.componentInstance;
  });

  afterEach(inject([OverlayContainer], (currentOverlayContainer: OverlayContainer) => {
    currentOverlayContainer.ngOnDestroy();
    overlayContainer.ngOnDestroy();
  }));

  it('should create and match snapshot', () => {
    fixture.detectChanges();
    expect(testComponent).toBeTruthy();
    expect(testComponent.comboBoxComponent).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });

  it('should render with default class name and ngClass', () => {
    fixture.detectChanges();
    nativeElement = fixture.nativeElement;
    const combobox = nativeElement.querySelector('md-combo-box');
    expect(combobox).toBeTruthy();
    expect(combobox.className).toContain('md-combo-box');
    expect(combobox.className).toContain('test');
  });

  it('should display search icon by default', () => {
    fixture.detectChanges();
    nativeElement = fixture.nativeElement;
    expect(nativeElement.querySelector('md-search-input')).toBeTruthy();
  });

  it('should not show searchIcon when hasSearchIcon is false', () => {
    testComponent.hasSearchIcon = false;
    fixture.detectChanges();
    nativeElement = fixture.nativeElement;
    expect(nativeElement.querySelector('md-input')).toBeTruthy();
  });

  it('should show options when search string is does exists"', fakeAsync(() => {
    testComponent.value = 'a';
    testComponent.comboBoxComponent.allowMutationObserver = false;
    fixture.detectChanges();
    testComponent.comboBoxComponent.handleInput();
    fixture.detectChanges();
    tick(500);
    const listItems = overlayContainerElement.querySelectorAll('.md-list-item');
    expect(listItems.length).toEqual(3);
  }));

  it('should not show any options when search string is does not exists"', fakeAsync(() => {
    testComponent.value = 'xyz';
    testComponent.comboBoxComponent.allowMutationObserver = false;
    fixture.detectChanges();
    testComponent.comboBoxComponent.handleInput();
    fixture.detectChanges();
    tick(500);
    const listItems = overlayContainerElement.querySelectorAll('.md-list-item');
    expect(listItems.length).toEqual(0);
  }));

  it('should close the options list when an option is selected"', fakeAsync(() => {
    testComponent.value = 'a';
    testComponent.comboBoxComponent.allowMutationObserver = false;
    fixture.detectChanges();
    testComponent.comboBoxComponent.handleInput();
    fixture.detectChanges();
    tick(500);
    let listItems = overlayContainerElement.querySelectorAll('.md-list-item');
    listItems[0].dispatchEvent(new MouseEvent('click'));
    fixture.detectChanges();
    tick(500);
    listItems = overlayContainerElement.querySelectorAll('.md-list-item');
    expect(listItems.length).toEqual(0);

    // click back on input tag, should open the options list once again
    testComponent.comboBoxComponent.handleInput();
    fixture.detectChanges();
    tick(500);
    listItems = overlayContainerElement.querySelectorAll('.md-list-item');
    expect(listItems.length).toEqual(3);
  }));

  it('should handle keyBoard events"', fakeAsync(() => {
    testComponent.value = 'a';
    testComponent.comboBoxComponent.allowMutationObserver = false;
    fixture.detectChanges();
    testComponent.comboBoxComponent.handleInput();
    fixture.detectChanges();
    tick(500);
    nativeElement = fixture.nativeElement;
    let input = nativeElement.querySelector('md-search-input');
    input.dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowDown',
      })
    );
    fixture.detectChanges();
    let listItems = overlayContainerElement.querySelectorAll('.md-list-item');
    expect(listItems[0].className).toContain('active');

    input.dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowDown',
      })
    );
    fixture.detectChanges();
    listItems = overlayContainerElement.querySelectorAll('.md-list-item');
    expect(listItems[1].className).toContain('active');

    input.dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'Enter',
      })
    );
    fixture.detectChanges();
    tick(500);
    input = nativeElement.querySelector('input');
    expect(testComponent.value).toEqual('ab');
  }));

  it('should handle keyBoard events with few items disabled"', fakeAsync(() => {
    testComponent.options = [
      {
        label: 'a',
        disabled: true,
      },
      {
        label: 'ab',
      },
      {
        label: 'abc',
      },
    ];
    testComponent.value = 'a';
    testComponent.comboBoxComponent.allowMutationObserver = false;
    fixture.detectChanges();
    testComponent.comboBoxComponent.handleInput();
    fixture.detectChanges();
    tick(500);
    nativeElement = fixture.nativeElement;
    const input = nativeElement.querySelector('md-search-input');
    input.dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowDown',
      })
    );
    fixture.detectChanges();
    let listItems = overlayContainerElement.querySelectorAll('.md-list-item');
    expect(listItems[1].className).toContain('active');

    input.dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowDown',
      })
    );
    fixture.detectChanges();
    listItems = overlayContainerElement.querySelectorAll('.md-list-item');
    expect(listItems[2].className).toContain('active');

    input.dispatchEvent(
      new KeyboardEvent('keydown', {
        code: 'ArrowDown',
      })
    );
    fixture.detectChanges();
    listItems = overlayContainerElement.querySelectorAll('.md-list-item');
    expect(listItems[1].className).toContain('active');

  }));

  it('on click outside, should close the options list"', fakeAsync(() => {
    testComponent.value = 'a';
    testComponent.comboBoxComponent.allowMutationObserver = false;
    fixture.detectChanges();
    testComponent.comboBoxComponent.handleInput();
    fixture.detectChanges();
    tick(500);
    const listItems = overlayContainerElement.querySelectorAll('.md-list-item');
    expect(listItems.length).toEqual(3);

    document.dispatchEvent(new MouseEvent('mousedown'));
    fixture.detectChanges();
    tick(500);

    expect(overlayContainerElement.textContent).toBe('');
  }));

});
