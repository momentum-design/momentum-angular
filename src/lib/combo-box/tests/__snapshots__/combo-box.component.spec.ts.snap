// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboBoxComponent should create and match snapshot 1`] = `
<test-combo-box
  clear="false"
  comboBoxComponent={[Function ComboBoxComponent]}
  hasSearchIcon={[Function Boolean]}
  options={[Function Array]}
  value=""
>
  <md-combo-box
    aria-haspopup="listbox"
    class="md-combo-box test ng-untouched ng-pristine ng-valid"
    ng-reflect-clear="false"
    ng-reflect-has-search-icon="true"
    ng-reflect-model=""
    ng-reflect-ng-class="test"
    ng-reflect-options="a,ab,abc"
    ng-reflect-value=""
    role="combobox"
  >
    <div
      cdk-overlay-origin=""
    >
      
      <md-search-input
        aria-autocomplete="list"
        class="ng-untouched ng-pristine ng-valid"
        ng-reflect-clear="false"
        ng-reflect-disabled="false"
        ng-reflect-is-disabled="false"
        ng-reflect-model=""
        ng-reflect-placeholder=""
        ng-reflect-theme=""
      >
        <div
          class="md-input-group md-search-input"
          ng-reflect-klass="md-input-group md-search-input"
          ng-reflect-ng-class="[object Object]"
        >
          
          <div
            class="md-input__icon-container"
          >
            <input
              class="md-input ng-untouched ng-pristine ng-valid"
              id=""
              ng-reflect-klass="md-input"
              ng-reflect-name=""
              ng-reflect-ng-class="[object Object]"
              placeholder=""
              type="text"
            />
            <i
              class="md-icon icon md-search-input__icon icon-search_20"
              ng-reflect-klass="md-icon icon md-search-input__"
              ng-reflect-ng-class="[object Object]"
            />
            
          </div>
        </div>
      </md-search-input>
      
    </div>
    
  </md-combo-box>
</test-combo-box>
`;
